### DISCUSSION TITLE
why use duckdb_temp_storage-index.tmp to make tmp file without mktemp to make random temp name

### DISCUSSION THREAD

**Discussion by wuyouying** [duckdb/duckdb#15156] [Q&A]
```
the code to make tmp file:
path(FileSystem::GetFileSystem(db).JoinPath(
	                                     temp_directory, "duckdb_temp_storage-" + to_string(index) + ".tmp")

It uses index as temp file'name,when I open two duckdbs and set the same temp_directory,
they will return error as:
IO Error: Could not read enough bytes from file "/tmp/duckdb_dp.data.tmp/duckdb_temp_storage-0.tmp": attempted to read 262144 bytes from location 307494912
```
Why not use mktemp to make random temp name,and thus avoid the error above?
Does it have plan to modify the porformance?

**Comment by Tishj** [comment]
> set the same temp_directory

Can you explain why you would do this?
I feel like you're asking for additional complexity in the system to solve a problem that doesn't need fixing

A unique temp folder is used by default

**Reply by wuyouying** [reply]
Now that create temp file and delete temp file is manual,but if  I kill the thread before DeleteTemporaryFile(id),the temp files will remain in disk.
```
unique_ptr<FileBuffer> StandardBufferManager::ReadTemporaryBuffer(MemoryTag tag, block_id_t id,
                                                                  unique_ptr<FileBuffer> reusable_buffer) {
	D_ASSERT(!temporary_directory.path.empty());
	D_ASSERT(temporary_directory.handle.get());
	if (temporary_directory.handle->GetTempFile().HasTemporaryBuffer(id)) {
		evicted_data_per_tag[uint8_t(tag)] -= Storage::BLOCK_SIZE;
		return temporary_directory.handle->GetTempFile().ReadTemporaryBuffer(id, std::move(reusable_buffer));
	}
	idx_t block_size;
	// open the temporary file and read the size
	auto path = GetTemporaryPath(id);
	auto &fs = FileSystem::GetFileSystem(db);
	auto handle = fs.OpenFile(path, FileFlags::FILE_FLAGS_READ);
	handle->Read(&block_size, sizeof(idx_t), 0);
	evicted_data_per_tag[uint8_t(tag)] -= block_size;

	// now allocate a buffer of this size and read the data into that buffer
	auto buffer = ReadTemporaryBufferInternal(*this, *handle, sizeof(idx_t), block_size, std::move(reusable_buffer));

	handle.reset(); _**// If I kill thread before DeleteTemporaryFile(id),the temp file will remain in disk.**_
	DeleteTemporaryFile(id); 
	return buffer;
}
```
So I think it may use something like mktemp to delete the temp file automatic when kill thread before DeleteTemporaryFile(id).
And so that it can resolve the error above at the same time.

**Reply by Tishj** [reply]
There is no need for you to manually create the folder. If you don't, duckdb will manage its lifetime and clean it up

**Reply by soerenwolfers** [reply]
@Tishj The default duckdb settings are such that all duckdb instances use the same folder. I agree with the OP that given this is the default configuration, duckdb should itself avoid clashes of the files used. I have opened https://github.com/duckdb/duckdb/issues/15173 about this, hoping it's what the OP here actually meant.

**Reply by Tishj** [reply]
All in-memory instances use the same folder*

While I agree that is a problem, that is an important detail

**Reply by chestnutsj** [reply]
why not  use mktmp to create temp file ?   there will be no duplication of names, and there will be no residual temporary files when the program crashes.

**Reply by carlopi** [reply]
Thanks for raising, someone is looking at this, help is coming!

**Reply by Tishj** [reply]
> why not use mktmp to create temp file ? there will be no duplication of names, and there will be no residual temporary files when the program crashes.

Thanks for the input, duckdb is not a native python application, it's not feasible to use `mktemp`
